src/driver-asus.c

```diff
@@ -0,0 +1,516 @@
+	mapping = ratbag_device_data_asus_get_button_mapping(device->data);
+
+	// get current profile id
+	profile_id = 0;
+	profile_data = zalloc(sizeof(*profile_data));
+	error = asus_get_profile_data(device, profile_data);
+	if (!error) {
+		profile_id = profile_data->profile_id;
+		log_debug(device->ratbag, "Initial profile is %d\n", profile_id);
+		log_debug(
+			device->ratbag, "Primary version %02X.%02X.%02X\n",
+			profile_data->version_primary[0],
+			profile_data->version_primary[1],
+			profile_data->version_primary[2]);
+		log_debug(
+			device->ratbag, "Secondary version %02X.%02X.%02X\n",
+			profile_data->version_secondary[0],
+			profile_data->version_secondary[1],
+			profile_data->version_secondary[2]);
+	}
+	free(profile_data);
+	if (error)
+		return error;
+
+	// read ratbag profiles
+	ratbag_device_for_each_profile(device, profile) {
+		if (profile->index == profile_id) {  // profile is already selected
+			profile->is_active = true;
+
+		} else {  // switch profile
+			profile->is_active = false;
+
+			log_debug(device->ratbag, "Switching to profile %d\n", profile->index);
+			error = asus_set_profile(device, profile->index);
+			if (error)
+				return error;
+		}
+
+		// get buttons
+		log_debug(device->ratbag, "Loading buttons data\n");
+		button_data = zalloc(sizeof(*button_data));
+		error = asus_get_button_data(device, button_data);
+		if (!error) {
+			ratbag_profile_for_each_button(profile, button) {
+				asus_index = mapping[button->index] != -1 ? mapping[button->index] : ASUS_BUTTON_MAPPING_ORDER[button->index];
+				if (asus_index == -1) {
+					log_debug(device->ratbag, "No mapping for button %d\n", button->index);
+					continue;
+				}
+
+				// disabled
+				if (button_data->data.button[asus_index].action == ASUS_BUTTON_CODE_DISABLED) {
+					button->action.type = RATBAG_BUTTON_ACTION_TYPE_NONE;
+					continue;
+				}
+
+				// got action
+				switch(button_data->data.button[asus_index].type) {
+				case ASUS_BUTTON_ACTION_TYPE_KEY:
+					button->action.type = RATBAG_BUTTON_ACTION_TYPE_KEY;
+					// asus code to linux code
+					button->action.action.key.key = ASUS_KEY_MAPPING[button_data->data.button[asus_index].action];
+					break;
+
+				case ASUS_BUTTON_ACTION_TYPE_BUTTON:
+					asus_index_to = asus_find_button_index_by_code(button_data->data.button[asus_index].action);
+					if (asus_index_to >= 0) {
+						button->action.type = ASUS_BUTTON_MAPPING[asus_index_to].type;
+						if (ASUS_BUTTON_MAPPING[asus_index_to].type == RATBAG_BUTTON_ACTION_TYPE_BUTTON)
+							button->action.action.button = ASUS_BUTTON_MAPPING[asus_index_to].button;
+						else if (ASUS_BUTTON_MAPPING[asus_index_to].type == RATBAG_BUTTON_ACTION_TYPE_SPECIAL)
+							button->action.action.special = ASUS_BUTTON_MAPPING[asus_index_to].special;
+					} else {
+						log_debug(device->ratbag, "Unknown action code %02x\n", button_data->data.button[asus_index].action);
+					}
+					break;
```

Comment:
for this code, I really recommend using temporary variables so you dont need to use `foo->bar[bat->index].baz`. On higher optimisation those are free anyway, so having something like this makes the code easier to read:
```c
   struct foo *mapping = &ASUS_BUTTON_MAPPING[asus_index_to];
   if (mapping->type == ACTION_TYPE_BUTTON)
      button->action.type = mapping->type;